### PR TITLE
Create valid error response

### PR DESCRIPTION
**Pull request by kickster97**
Creates valid error responses even if there is an exception in `array_iterator_to_json` 
The fabricated responses are handled in the Table javascript class

### CODE REVIEW

File: src/lavinmq/http/controller.cr [hunk 1]
```diff
@@ -102,11 +102,15 @@ module LavinMQ
         size = 0
         total = 0
         json.array do
-          iterator.each_with_index do |o, i|
-            total += 1
-            next if i < start || start + page_size <= i
-            o.to_json(json)
-            size += 1
+          begin
```
Comment (baelter) [inline src/lavinmq/http/controller.cr:105]: Weird that `ameba` doesn't find this but you don't need a begin block, you can rescue in the outer block.

File: src/lavinmq/http/controller.cr [hunk 2]
```diff
@@ -102,11 +102,15 @@ module LavinMQ
...
+            iterator.each_with_index do |o, i|
+              total += 1
+              next if i < start || start + page_size <= i
+              o.to_json(json)
+              size += 1
+            end
+          rescue e
+            json.object { json.field "error", e.message }
```
Comment (baelter) [inline src/lavinmq/http/controller.cr:113]: Should we use the `{ error: type, reason: message }` format we use in the `ApiErrorHandler` ?
And set a different status code? 500?

Comment (kickster97) [reply src/lavinmq/http/controller.cr:113]: Yes, great!

Comment (carlhoerberg) [reply src/lavinmq/http/controller.cr:113]: i'm not sure about the response code actually... if the result set is very large, and say 900 items have already been written to the response, then a) is it a valid or invalid response? b) the response header with the status code might already have been written to the socket and can't be changed

File: src/lavinmq/http/controller.cr [hunk 3]
```diff
@@ -102,11 +102,15 @@ module LavinMQ
...
-          iterator.each_with_index do |o, i|
-            total += 1
-            next if i < start || start + page_size <= i
-            o.to_json(json)
-            size += 1
+          begin
+            iterator.each_with_index do |o, i|
+              total += 1
+              next if i < start || start + page_size <= i
+              o.to_json(json)
```
Comment (carlhoerberg) [inline src/lavinmq/http/controller.cr:109]: only wrap this line in the `begin..rescue` block.

**Review by carlhoerberg — CHANGES_REQUESTED** [review]

File: src/lavinmq/http/controller.cr (hunk 2 above)
Comment (baelter) [reply src/lavinmq/http/controller.cr:113]: True, like we fail to generate a response for one of 900 queues, this happens in rabbitmq as well, they just send some null fields back.

File: static/js/table.js [hunk 4]
```diff
@@ -170,7 +170,7 @@ function renderTable (id, options = {}, renderRow) {
     for (let i = 0; i < data.length; i++) {
       const item = data[i]
       const foundIndex = findIndex(t.rows, start, item)
-      if (foundIndex !== -1) {
+      if (foundIndex !== -1 && item.error === undefined) {
```
Comment (carlhoerberg) [inline static/js/table.js:173]: you want to check for the error property earlier, not when we've already tried to "find" it.

**Review by carlhoerberg — CHANGES_REQUESTED** [review]

File: src/lavinmq/http/controller.cr (hunk 2 above)
Comment (kickster97) [reply src/lavinmq/http/controller.cr]: So maybe we should just filter out the failure from items, and still render the queues with valid responses? (with a warning that we failed to load some data) ?

File: src/lavinmq/http/controller.cr [hunk 5]
```diff
@@ -108,6 +108,8 @@ module LavinMQ
             o.to_json(json)
             size += 1
           end
+        rescue e
+          json.object { json.field "error", e.message }
```
Comment (carlhoerberg) [inline src/lavinmq/http/controller.cr:112]:
```suggestion
            begin
              o.to_json(json)
            rescue e
              json.object { json.field "error", e.message }
            end
            size += 1
          end
```

File: static/js/table.js [hunk 6]
```diff
@@ -178,7 +180,6 @@ function renderTable (id, options = {}, renderRow) {
             renderRow(t.rows[foundIndex], data[foundIndex], true)
             setKeyAttributes(t.rows[foundIndex], data[foundIndex])
           }
-        } else {
```
Comment (carlhoerberg) [inline static/js/table.js:181]: Are you sure about this? I don't think this line should be remove.

File: static/js/table.js [hunk 7]
```diff
@@ -131,6 +131,8 @@ function renderTable (id, options = {}, renderRow) {
     const tableError = document.getElementById(id + '-error')
     return HTTP.request('GET', fullUrl).then(function (response) {
       tableError.textContent = ''
+      const hasError = response.items.some(i => i.error !== undefined)
+      if (hasError) { throw "item error: invalid item type" }
```
Comment (carlhoerberg) [inline static/js/table.js:135]: Can't we write something more specific? Like taking the first error message?

**Review by carlhoerberg — CHANGES_REQUESTED** [review]

**Comment by kickster97** [comment]
Rescue when mapping `details_tuple` to the `iterator` because we can have errors here which we will not catch if we wait untill `array_iterator_to_json` to rescue. 

Made error display a little nicer, a complete renovation on msg/error displaying for everything could be done as part of UI-project.

File: static/main.css [hunk 8]
```diff
@@ -381,7 +381,12 @@ footer .logo {
 }
 
 #table-error {
-  color: red;
+  display: none;
```
Comment (baelter) [inline static/main.css:384]: Other use of `#table-error` will be hidden now. We always have to change `display` when adding content to this element now. Maybe write a function in table that does it.

https://github.com/cloudamqp/lavinmq/blob/2bae422dc0ddcebf728fd7cde67f6c797e01a0d3/static/vhosts.html#L74
https://github.com/cloudamqp/lavinmq/blob/83efd006dab9cf727ce3cd44776c9216ce139d43/static/users.html#L89

Comment (kickster97) [reply static/main.css:384]: ah yes missed those, thanks!

File: static/js/table.js [hunk 9]
```diff
@@ -437,4 +443,14 @@ function updateQueryState (params) {
   window.history.replaceState(null, '', newurl)
 }
 
-export { renderCell, renderTable, renderHtmlCell }
+function toggleDisplayError (tableError, display, message) {
+  if (display) {
+    tableError.style.display = 'block'
+    tableError.textContent = 'Error fetching data: {' + message + '}'
```
[the remark this answers is not in the document]
Comment (kickster97) [reply static/js/table.js:449]: Ah right, maybe put "Something went wrong: " instead? 
Agree, will fix

File: static/js/table.js [hunk 10]
```diff
@@ -139,9 +138,9 @@ function renderTable (id, options = {}, renderRow) {
       updateTable(response)
     }).catch(function (e) {
       if (e.body) {
-        tableError.textContent = 'Error fetching data: ' + e.body
+        toggleDisplayError(id, true, e.body)
       } else {
-        tableError.textContent = "Error fetching data: Can't reach server, please try to refresh the page!"
+        toggleDisplayError(id, true, "{ error: Can't reach server, please try to refresh the page }")
```
Comment (baelter) [inline static/js/table.js:143]:
```suggestion
        toggleDisplayError("Error fetching data: Can't reach server, please try to refresh the page.")
```

File: static/js/table.js [hunk 11]
```diff
@@ -139,9 +138,9 @@ function renderTable (id, options = {}, renderRow) {
       updateTable(response)
     }).catch(function (e) {
       if (e.body) {
-        tableError.textContent = 'Error fetching data: ' + e.body
+        toggleDisplayError(id, true, e.body)
```
Comment (baelter) [inline static/js/table.js:141]:
```suggestion
        toggleDisplayError('Error fetching data: ' + e.body)
```

File: static/js/table.js [hunk 12]
```diff
@@ -437,4 +441,15 @@ function updateQueryState (params) {
   window.history.replaceState(null, '', newurl)
 }
 
-export { renderCell, renderTable, renderHtmlCell }
+function toggleDisplayError (id, display, message) {
+  const tableError = document.getElementById(id + '-error')
```
Comment (baelter) [inline static/js/table.js:445]:
```suggestion
  const tableError = document.getElementById('table-error')
```

File: static/users.html [hunk 13]
```diff
@@ -86,7 +88,7 @@ <h3>Add user</h3>
           Table.renderCell(tr, 3, hasPassword, 'center')
         })
       }).catch(e => {
-        document.getElementById("table-error").textContent = e.status === 401 ? "Access refused: You need administrator role to see this view" : e.body
+        Table.toggleDisplayError(document.getElementById('table-error'), true, e.status === 401 ? 'Access refused: You need administrator role to see this view' : e.body)
```
Comment (baelter) [inline static/users.html:91]:
```suggestion
        Table.toggleDisplayError(e.status === 401 ? 'Access refused: You need administrator role to see this view' : e.body)
```

File: static/vhosts.html [hunk 14]
```diff
@@ -63,16 +65,15 @@ <h3>Add virtual host</h3>
...
         const urlEncodedVhost = encodeURIComponent(item.name)
         const permissionsUrl = '/api/vhosts/' + urlEncodedVhost + '/permissions'
         const permissionsPromise = new Promise(function (resolve, reject) {
           HTTP.request('GET', permissionsUrl).then(permissions => {
             window.sessionStorage.setItem(permissionsUrl, JSON.stringify(permissions))
             resolve(permissions)
           }).catch(e => {
-              document.getElementById("table-error").textContent = e.status === 401 ? e.body + ": You need administrator role to see this view" : e.body
-              reject(e)
+            Table.toggleDisplayError(document.getElementById('table-error'), true, e.status === 401 ? e.body + ': You need administrator role to see this view' : e.body)
```
Comment (baelter) [inline static/vhosts.html:75]:
```suggestion
            Table.toggleDisplayError(e.status === 401 ? e.body + ': You need administrator role to see this view' : e.body)
```

File: static/js/table.js [hunk 15]
```diff
@@ -128,9 +128,8 @@ function renderTable (id, options = {}, renderRow) {
 
   function fetchAndUpdate () {
     const fullUrl = `${url}?${buildQuery(currentPage)}`
-    const tableError = document.getElementById(id + '-error')
     return HTTP.request('GET', fullUrl).then(function (response) {
-      tableError.textContent = ''
+      toggleDisplayError(id, false, '')
```
Comment (baelter) [inline static/js/table.js:132]:
```suggestion
      toggleDisplayError(false)
```

File: static/js/table.js [hunk 16]
```diff
@@ -167,26 +166,31 @@ function renderTable (id, options = {}, renderRow) {
...
+          start = i + 1
         }
-        start = Math.min(i + 1, foundIndex)
-      } else {
-        const tr = t.insertRow(i)
-        setKeyAttributes(tr, item)
-        renderRow(tr, item, true)
-        start = i + 1
+      } catch (e) {
+        toggleDisplayError(id, true, e)
```
Comment (baelter) [inline static/js/table.js:193]:
```suggestion
        toggleDisplayError(item.error || e.message)
```

**Review by baelter — COMMENTED** [review]
Some refactoring suggestions around the toggleError function

**Pull request merged**